deeplearning4j/deeplearning4j-ui-parent/deeplearning4j-play/src/main/java/org/deeplearning4j/ui/api/UIServer.java

```diff
@@ -38,9 +39,27 @@
      * @return UI instance for this JVM
      */
     public static synchronized UIServer getInstance() {
-        if (uiServer == null) {
-            PlayUIServer playUIServer = new PlayUIServer();
-            playUIServer.runMain(new String[] {"--uiPort", String.valueOf(PlayUIServer.DEFAULT_UI_PORT)});
+        return getInstance(false, null);
+    }
+
+    /**
+     * Get (and, initialize if necessary) the UI server.
+     * Singleton pattern - all calls to getInstance() will return the same UI instance.
+     * @param multiSession in multi-session mode, multiple training sessions can be visualized in separate browser tabs.
+     *                     <br/>URL path will include session ID as a parameter, i.e.: /train becomes /train/:sessionId
+     * @param statsStorageProvider function that returns a StatsStorage containing the given session ID.
+     *                             <br/>Use this to auto-attach StatsStorage if an unknown session ID is passed
+     *                             as URL path parameter in multi-session mode, or leave it {@code null}.
+     * @return UI instance for this JVM
+     */
+    public static synchronized UIServer getInstance(boolean multiSession, Function<String, StatsStorage> statsStorageProvider) {
+        if (uiServer== null) {
+            PlayUIServer playUIServer = new PlayUIServer(PlayUIServer.DEFAULT_UI_PORT, multiSession);
+            playUIServer.setMultiSession(multiSession);
+            if (statsStorageProvider != null) {
+                playUIServer.autoAttachStatsStorageBySessionId(statsStorageProvider);
+            }
+            playUIServer.runMain(new String[] {});
             uiServer = playUIServer;
         }
         return uiServer;
```

Comment:
Maybe it's unnecessary, but one way I can think of breaking this is doing the following:
```
UIServer.getInstance();
//some time later...
UIServer.getInstance(true, null);    //User expects multi-session, but it's already initialized as single session mode.
```
Perhaps validating the already-initialized UIServer is actually multi-session would be a good idea?

Reply:
Yes, i this case, user would get a single-session instance. I would add user-side validation, by introducing `public abstract boolean isMultiSession` in `UIServer`, and a warning message in `getInstance(true, null)`, if the instance is not multi-session.

Reply:
Instead of a warning message, I would throw a `RuntimeException`. This way, we only return an instance, if the call to the singleton factory method is consistent with the first call. Is `RuntimeException` the right exception to throw in this case?

Reply:
Yeah, I think I'd prefer an exception in this case... it's clear, unambiguous and won't be missed my the user. The exception text should tell user what's going on, why, and what they can do to fix it (I assume just stopping the UI server instance, and starting a new one should suffice).
An isMultiSession method in addition (and used in exception) sounds fine too.

Reply:
Well, I have coded the validation with an exception. I have made two test cases to demonstrate how it works. I will add your solution suggestion 
> stopping the UI server instance, and starting a new one

to the exception text. 

On the other hand, I have an issue with stopping the server: I tried to run UIServer.stop() in one of the new tests in `TestPlayUIMultiSession`, but it threw this exception:
```
java.lang.NoClassDefFoundError: org/slf4j/bridge/SLF4JBridgeHandler
	at play.api.Logger$.shutdown(Logger.scala:306)
	at play.core.server.Server$class.stop(Server.scala:69)
	at play.core.server.NettyServer.stop(NettyServer.scala:175)
	at play.server.Server.stop(Server.java:28)
	at org.deeplearning4j.ui.play.PlayUIServer.stop(PlayUIServer.java:422)
```
See [this gist](https://gist.github.com/printomi/c6999cfe57426d77630fb60b00ecf803) for the full exception stack trace. 
Do I miss a dependency in pom.xml?

Reply:
@AlexDBlack I have reported the above mentioned issue with stopping UIServer #7274 . If there is an easy solution, that could be implemented in this PR. What do you think?

Reply:
I think, that commits https://github.com/deeplearning4j/deeplearning4j/pull/7185/commits/67afb807c6453b27163e91628d82beee969e15cf and https://github.com/deeplearning4j/deeplearning4j/pull/7185/commits/dad7bb9da0c65be1befe0affd1968f1b66137549 solve the issue. @AlexDBlack please, resolve this conversation, if you accept the solution.